File: python/P402.py
class Solution(object):
    def removeKdigits(self, num, k):
        out = []
        for each in num:
            while k and out and out[-1] > each:
                out.pop()
                k -= 1
            out.append(each)
        return ''.join(out[:-k or None]).lstrip('0') or '0' #[:-k] will reach number kth element

File: java/P402.java
class Solution {
    public String removeKdigits(String num, int k) {
        if (k == num.length()) return "0";
        Stack<Character> stack = new Stack<>();
        int i = 0;
        while (i < num.length()) {
            while (k >0 && !stack.isEmpty() && stack.peek() > num.charAt(i)) {
                k--;
                stack.pop();
            }
            stack.push(num.charAt(i));
            i++;
        }
        while (k>0) {
            stack.pop();
            k--;
        }        
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        sb.reverse();
        while (sb.length()>1 && sb.charAt(0) == '0') {
            sb.deleteCharAt(0);
        }
        return sb.toString();
    }
}